Add 'adb_response' attribute to Android TV / Fire TV

JeffLIrion opened the pull request:
## Description:

This stores the output, if any, of the `androidtv.adb_command` service in a state attribute -- `'adb_response'` -- so that it can be used in automations / scripts. 

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

JeffLIrion committed:
Add 'adb_response' attribute to Android TV / Fire TV

rytilahti commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -304,12 +310,17 @@ def adb_command(self, cmd):
         if key:
-            return self.aftv.adb_shell('input keyevent {}'.format(key))
+            self.aftv.adb_shell('input keyevent {}'.format(key))
+            self._adb_response = ''
```
Is there no response for this (for example, if the call succeeded or not)?

JeffLIrion replied:
Right, there's no response.

rytilahti commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -304,12 +310,17 @@ def adb_command(self, cmd):
 
-        return self.aftv.adb_shell(cmd)
+        response = self.aftv.adb_shell(cmd)
+        self._adb_response = str(response) if response else ''
```
Is the else part for error reporting? Should errors also be reported somehow?

JeffLIrion replied:
`response` is either a string or `None`, so I'm handling both cases.

rytilahti commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -157,9 +157,9 @@ def service_adb_command(service):
             output = target_device.adb_command(cmd)
 
             # log the output if there is any
-            if output and (not isinstance(output, str) or output.strip()):
```
Is this change just code cleanup, or can it change the behavior?

JeffLIrion replied:
It won't change the behavior. In the `adb_command` I made sure that the return value is a string, so I simplified the logic here.

rytilahti commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -157,9 +157,9 @@ def service_adb_command(service):
+            if output.strip():
                 _LOGGER.info("Output of command '%s' from '%s': %s",
-                             cmd, target_device.entity_id, repr(output))
+                             cmd, target_device.entity_id, output)
```
If repr is really wanted, you can use `%r` as the format string, afaik.

JeffLIrion replied:
I think `%s` is fine. In fact, I don't think it makes any difference, since `output` will only be logged when it is a non-empty string.

rytilahti commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -224,6 +224,7 @@ def __init__(self, aftv, name, apps, turn_on_command,
             self.exceptions = (ConnectionResetError, RuntimeError)
 
         # Property attributes
+        self._adb_response = ''
```
Is there some reason to use empty string instead of `None`?

JeffLIrion replied:
I could use `None` instead, if that would be preferable.

JeffLIrion replied:
I switched to using `None` instead of an empty string. So if there wasn't any output, the states page will report that the `adb_response` attribute is `null`.

JeffLIrion committed:
Use None instead of empty string for empty ADB responses

JeffLIrion commented:
It can be the case that this `adb_response` attribute is really long (for example, the response from `dumpsys power`). But (a) I don't think this hurts anything and (b) this is a more advanced feature anyways for users who want to send an ADB command and parse its output, so it's up to the user to send reasonable commands and manage the output.

JeffLIrion committed:
Initialize self._adb_response as None, not empty string

JeffLIrion committed:
Update the state after sending an ADB command

This ensures that the `'adb_response'` attribute contains the response to the latest command

JeffLIrion commented:
@rytilahti are there any other changes you'd like me to make?

rytilahti commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -304,12 +310,24 @@ def adb_command(self, cmd):
+            return self._adb_response
+
+        response = self.aftv.adb_shell(cmd)
+        if isinstance(response, str) and response.strip():
```
This feels a bit hacky.. Is there some case where the adb_shell() call would be returning something else than a string? If not, it'd be just easier to check for that here? Otherwise I think this would be better handled in the backend lib.

Other than that it looks fine to me; I don't have a device at hand, so I cannot really do any testing on functionality.

JeffLIrion replied:
It's either a string or `None`. I've been using this version of the component, it works as expected.

rytilahti replied:
I think I'd personally prefer then checking against Noneness here, but it's fine as it is. Let's merge this, thanks! :tada:
